pywbem_mock/_namespaceprovider.py

```diff
@@ -0,0 +1,290 @@
+    the `namespace` as the value of the name property in the instance. The
+    instance is created in the `interop_namespace`
+
+    Parameters:
+
+      namespace
+        Namespace that this instance defines
+
+      interop_namespace
+        Interop namespace for this environment.
+
+      klass
+
+    Raises:
+
+        CIMERROR: For errors encountered with CreateInstance
+
+    """
+    properties = NocaseDict(
+        [('Name', namespace),
+         ('CreationClassName', klass.classname),
+         ('ObjectManagerName', OBJECTMANAGERNAME),
+         ('ObjectManagerCreationClassName', 'CIM_ObjectManager'),
+         ('SystemName', SYSTEMNAME),
+         ('SystemCreationClassName', SYSTEMCREATIONCLASSNAME)])
+
+    new_instance = CIMInstance.from_class(klass, namespace, properties)
+    conn.CreateInstance(new_instance, interop_namespace)
+
+
+def create_cimnamespace_instances(conn, interop_namespace):
+    """
+    Create an instance of CIM_Namespace for every existing namespace
+
+    """
+    insts = conn.EnumerateInstances('CIM_Namespace',
+                                    namespace=interop_namespace,
+                                    LocalOnly=False,
+                                    DeepInheritance=False)
+    ns_dict = NocaseDict([(inst.name, None) for inst in insts])
+
+    # Get the current namespaces directly from the repository since the
+    # instances of CIM_Namespace not yet set up.
+    namespaces = conn.mainprovider.namespaces
+
+    classname = 'CIM_Namespace'
+    klass = conn.GetClass(classname, interop_namespace, LocalOnly=False,
+                          IncludeQualifiers=True,
+                          IncludeClassOrigin=True)
+
+    for ns in namespaces:
+        if ns not in ns_dict:
+            create_cimnamespace_instance(conn, ns, interop_namespace, klass)
+
+
+class CIMNamespaceProvider(InstanceWriteProvider):
+    """
+    Implements the user defined provider for the class CIM_Namespace.
+
+    This provider provides the create, modify, and delete methods for
+    adding an instance of CIM_Namspace when a namespace is created or
+    deleted in a pywbem mock environment.
+
+    This class and the instances of this class only exist in the interop
+    namespace for a server.
+    """
+    # these class level attributes must exist to define the provider
+    provider_type = 'instance'
+    provider_classnames = 'CIM_Namespace'
+
+    def __init__(self, cimrepository):
+        """
+        Set up the cim repository
+
+        """
+        super(CIMNamespaceProvider, self).__init__(cimrepository)
+
+        # Set by install_provider since __init__ does not have access to
+        # conn
+        self.interop_namespaces = None
+
+    def __repr__(self):
+        return _format(
+            "CIMNamespaceProvider("
+            "provider_type={s.provider_type}, "
+            "provider_classnames={s.provider_classnames})",
+            s=self)
```

Comment:
I think that repr is not necessary since you don't add anything.

I suggest to change the base class repr so it shows the self object's class name (`self.__class__.__name__`)

Reply:
TODO

Reply:
Lets move the provider_type into the superclass.  TODO